src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -939,16 +864,22 @@ ModifyQuerySupported(Query *queryTree, Query *originalQuery, bool multiShardQuer
 			/* for other kinds of relations, check if its distributed */
 			else
 			{
-				Oid relationId = rangeTableEntry->relid;
-
-				if (!IsCitusTable(relationId))
+				if (IsRelationLocalTableOrMatView(rangeTableEntry->relid) &&
+					containsLocalTableDistributedTableJoin)
 				{
 					StringInfo errorMessage = makeStringInfo();
 					char *relationName = get_rel_name(rangeTableEntry->relid);
-
-					appendStringInfo(errorMessage, "relation %s is not distributed",
-									 relationName);
-
+					if (IsCitusTable(rangeTableEntry->relid))
+					{
+						appendStringInfo(errorMessage,
+										 "citus local table %s cannot be joined with these distributed tables",
```

Comment:
what does these refer to?

Reply:
This is not a good error message.

This means `citus local table is involved in a join with a distributed table`. SELECT * FROM citus_local join dist...

Since this is in router planner, should we say something like: "router planner cannot handle citus local table joins with distributed tables"

The `router planner` part can be confusing for users. so maybe "Citus local tables cannot be joined with distributed tables".

I guess none of them is solution oriented though, we could mention the usage of CTE here maybe.

Reply:
Another thing is that we should probably not expose `citus local` error messages to the user, but that can go away when we hide it from the user in this cycle since we need a consensus on how to hide it in errors.